### PR TITLE
Remove legacy C++ memory optimization codes

### PR DESCRIPTION
**Pull request by sneaxiy**
As title.

This PR also enables `build_strategy.memory_optimize` when gc is disabled by default.

We have refactored in-place and memory-optimize passes in #17911 and #18735, which would never rename variable names. We do not need the legacy strategies any more.

### CODE REVIEW

File: paddle/fluid/framework/parallel_executor.cc [hunk 1]
```diff
@@ -252,7 +252,21 @@ ir::Graph *ParallelExecutorPrivate::ApplyMemoryOptimizePass(ir::Graph *graph) {
     VLOG(10) << "buffer_shared_inplace_pass Applied";
   }
 
-  if (build_strategy_.memory_optimize_) {
+  /**
```
Comment (zhhsplendid) [inline paddle/fluid/framework/parallel_executor.cc:255]: The comment logic doesn't match the code. In the comment you said 
"If gc is enabled, BuildStrategy.memory_optimize = False.
If gc is disabled, BuildStrategy.memory_optimize = True.
...
gc + memory_optimize is worse than gc only."

So your code logic should probably be without "if (!build_strategy_.memory_optimize_)" 
The if condition makes it possible for user to set both
build_strategy_.memory_optimize_ and gc to True.

Please consider to modify the comment or the code.

File: paddle/fluid/framework/parallel_executor.cc [hunk 2]
```diff
@@ -313,6 +327,9 @@ ir::Graph *ParallelExecutorPrivate::ApplyMemoryOptimizePass(ir::Graph *graph) {
     eager_deletion_pass->SetNotOwned(ir::kAllPlaces, &places_);
     graph = eager_deletion_pass->Apply(graph);
     VLOG(10) << "EagerDeletionPass Applied";
+    LOG(INFO) << "Garbage collection strategy is enabled, when "
+              << "FLAGS_eager_delete_tensor_gb="
```
Comment (zhhsplendid) [inline paddle/fluid/framework/parallel_executor.cc]: If you have time, add places around '='

such as "FLAGS_eager_delete_tensor_gb = "

File: paddle/fluid/framework/parallel_executor.cc [hunk 3]
```diff
@@ -313,6 +327,9 @@ ir::Graph *ParallelExecutorPrivate::ApplyMemoryOptimizePass(ir::Graph *graph) {
...
+              << GetEagerDeletionThreshold() / (static_cast<int64_t>(1) << 30);
```
Comment (zhhsplendid) [inline paddle/fluid/framework/parallel_executor.cc]: Can we write like this?
<< (GetEagerDeletionThreshold() >> 30);

File: paddle/fluid/framework/parallel_executor.cc [hunk 4]
```diff
@@ -252,7 +252,21 @@ ir::Graph *ParallelExecutorPrivate::ApplyMemoryOptimizePass(ir::Graph *graph) {
     VLOG(10) << "buffer_shared_inplace_pass Applied";
   }
 
-  if (build_strategy_.memory_optimize_) {
+  /**
+   * NOTE(zjl): If BuildStrategy.memory_optimize = None in Python,
```
Comment (luotao1) [inline paddle/fluid/framework/parallel_executor.cc]: `NOTE(zjl)`->`TODO(zengjinle)`? It's easier to find who write these with the full name.

**Review by luotao1 — COMMENTED** [review]
LGTM for remove the codes.

File: paddle/fluid/framework/parallel_executor.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/framework/parallel_executor.cc:255]: Discussed offline. User can forcely turn on `memory_optimize` by setting True. The comment is right.

File: paddle/fluid/framework/parallel_executor.cc (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/framework/parallel_executor.cc]: Done.

File: paddle/fluid/framework/parallel_executor.cc (hunk 3 above)
Comment (sneaxiy) [reply paddle/fluid/framework/parallel_executor.cc]: Done.

File: paddle/fluid/framework/parallel_executor.cc (hunk 4 above)
Comment (sneaxiy) [reply paddle/fluid/framework/parallel_executor.cc]: Done.

**Review by lanxianghit — APPROVED** [review]

**Review by xsrobin — APPROVED** [review]

**Pull request merged**